homeassistant/components/environment_canada/sensor.py

```diff
@@ -49,107 +51,98 @@ def validate_station(station):
 )
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+async def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
     """Set up the Environment Canada sensor."""
 
     if config.get(CONF_STATION):
-        ec_data = ECData(
+        ec_data = ECWeather(
             station_id=config[CONF_STATION], language=config.get(CONF_LANGUAGE)
         )
     else:
         lat = config.get(CONF_LATITUDE, hass.config.latitude)
         lon = config.get(CONF_LONGITUDE, hass.config.longitude)
-        ec_data = ECData(coordinates=(lat, lon), language=config.get(CONF_LANGUAGE))
+        ec_data = ECWeather(coordinates=(lat, lon), language=config.get(CONF_LANGUAGE))
+
+    async def async_update_data():
+        """Fetch data from Environment Canada."""
+        async with async_timeout.timeout(10):
+            await ec_data.update()
+        ec_data.conditions.update(ec_data.alerts)
+        return ec_data.conditions
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="environment_canada_sensor",
+        update_method=async_update_data,
+        update_interval=timedelta(minutes=5),
+    )
 
-    sensor_list = list(ec_data.conditions) + list(ec_data.alerts)
-    add_entities([ECSensor(sensor_type, ec_data) for sensor_type in sensor_list], True)
+    # Fetch initial data so we have data when entities subscribe
+    await coordinator.async_refresh()
 
+    async_add_devices(
+        ECSensor(coordinator, sensor_type, ec_data.metadata)
+        for sensor_type in coordinator.data.keys()
+    )
 
-class ECSensor(Entity):
+
+class ECSensor(CoordinatorEntity, Entity):
```

Comment:
We don't need to inherit from `Entity` when inheriting `CoordinatorEntity`. The former is the common parent.